modin/pandas/test/dataframe/test_join_sort.py

```diff
@@ -158,6 +158,100 @@ def test_join(test_data, test_data2):
         df_equals(modin_join, pandas_join)
 
 
+def test_join_4427():
+    count_cols = 16
+    data = np.random.uniform(0, 100, size=(1_000, count_cols))
+    left_columns = [f"lcol{x}" for x in range(count_cols)]
+    right_columns = [f"rcol{x}" for x in range(count_cols)]
+    modin_df1 = pd.DataFrame(data, columns=left_columns)
+    modin_df2 = pd.DataFrame(data, columns=right_columns)
+
+    pandas_df1 = pandas.DataFrame(data, columns=left_columns)
+    pandas_df2 = pandas.DataFrame(data, columns=right_columns)
+
+    for on in (None, left_columns[0], left_columns[:2]):
+        # check general `join` code branch
+        eval_general(
+            modin_df1,
+            pandas_df1,
+            lambda df: df.join(
+                modin_df2.set_index(right_columns[:2])
+                if isinstance(df, pd.DataFrame)
+                else pandas_df2.set_index(right_columns[:2]),
+                on=on,
+            ),
+        )
```

Comment:
maybe something like
```python
def _get_other(df):
    return modin_df2 if isinstance(df, pd.DataFrame) else pandas_df2

eval_general(
    modin_df1,
    pandas_df1,
    lambda df, other: df.join(other, on=on),
    other=_get_other,
)
```
(and use this `other=_get_other` in all `eval_general()`?)